### PR TITLE
Manual unity

### PR DESCRIPTION
**Pull request by drwells**
I am not sure if we want to actually use this but I figured I would throw it out there. Regardless, there is a bunch of hacky stuff in the history (such as the original version of #4422 and some other patches to get around name clashes). I added a few more unity files since yesterday: the final version requires roughly 25% less time to compile than master. Not bad!

with unity:
```
        Command being timed: "make -j10"
        User time (seconds): 7483.28
        System time (seconds): 178.97
        Percent of CPU this job got: 757%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 16:52.12
```

on master:
```
        Command being timed: "make -j10"
        User time (seconds): 9695.31
        System time (seconds): 271.92
        Percent of CPU this job got: 762%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 21:47.83
```

Essentially, all I did was bolt together, manually, all of the files that took GCC less than 15 seconds to compile (which is most files in the library). Everything in `base` got combined: almost everything in `lac`, `grid`, and `fe` also got put together.

### CODE REVIEW

**Comment by tamiko** [comment]
@drwells 
I have the feeling that using cotire with #4441 achieves comparable speedup (if you issue something like `make all_unity`, or `make obj_XXX_(debug|release)_unity` with far greater flexibility and a much less intrusive patch set.

**Comment by tamiko** [comment]
That said - I would like to merge at least parts of this pull request that fix up compilation units. (The cotire build fails spectacularly at the moment, details on #4441).

**Comment by drwells** [comment]
I cleaned things up a lot. This PR currently duplicates the commit in #4618 so that it compiles correctly.

The unity build implementation does not need to do much:
1. We gained a new option, `DEAL_II_UNITY_BUILD`, to pass to cmake. If this is set to `ON` (the default is `OFF`) then we do the unity build.
2. Most `CMakeLists.txt` files in the `source` directory now contain two lists of targets: the ones that should be concatenated together and the ones that should not.
3. CMake generates 'unity' files that `#include` together the aforementioned files.
4. CMake adds the current directory to the included file path and the unity files to the list of build targets.

The main problems with this are that it is somewhat invasive and we need to manually maintain the list of files that are too large to be included in the unity files (e.g., `matrix_creator_inst2.cc.o` takes 61 seconds already; this is about the time required by most of the unity files).

I suspect that the manual intervention will improve performance over cotire but I have not yet measured cotire.

The things left at this point are:
[ ] - merge #4618 
[ ] - add more documentation regarding how 'unity build' is defined and how to do it
[ ] - clean up comments and the unity macro

**Comment by drwells** [comment]
ping @tamiko this is the designed I described on Thursday.

**Comment by drwells** [comment]
I reran the unity build on the 16 core machine to get some more profiling results. It still saves a fair amount of time but the improvement is not quite as large as when I used a mere 4 core machine:

unity time usage:
```
make -j32  10298.66s user 332.00s system 2671% cpu 6:37.86 total
```

unity space usage:
```
[wellsd2@rtg1 build]$ du -h --max-depth=1 ./
28K     ./examples
120K    ./bin
4.0K    ./temp
32K     ./contrib
3.2M    ./tests
300K    ./cmake
3.4G    ./source
79M     ./bundled
1.3M    ./CMakeFiles
1.2M    ./Testing
1.9G    ./lib
112K    ./share
60K     ./doc
52K     ./include
5.3G    ./
```

master time usage:
```
make -j32  11862.98s user 474.99s system 2769% cpu 7:25.48 total
```

master space usage:
```
[wellsd2@rtg1 build]$ du -h --max-depth=1 ./
28K     ./examples
120K    ./bin
4.0K    ./temp
32K     ./contrib
3.2M    ./tests
300K    ./cmake
3.9G    ./source
79M     ./bundled
1.3M    ./CMakeFiles
1.2M    ./Testing
2.0G    ./lib
112K    ./share
60K     ./doc
52K     ./include
5.9G    ./
```

It looks like the main advantage is that we don't have to compile 600 megabytes of extra material: a fair amount of savings.

**Comment by drwells** [comment]
This is now ready for review.

**Comment by bangerth** [comment]
One might surmise that the unity build also opens up possibilities for optimization that aren't there when building each file separately. If you happen to have both builds still around, can you run, say, step-6 with a few more refinements and in optimized mode and see whether it makes a difference?

**Comment by drwells** [comment]
I reran step-6 a lot of times and the noise in the profiling seems larger than any difference I can measure.

**Comment by bangerth** [comment]
OK, thanks for testing.

**Comment by kronbichler** [comment]
I like this as it saves both time and a bit of disk space. Just a side question regarding the sizes of `src` and `lib`: Did you enable debug symbols also for the release build? It appears to me that `2.0G` for `lib` is rather large as I have only `1.2G` with everything enabled except for OpenCASCADE.

**Comment by drwells** [comment]
Yes; I did compile in both debug and release mode with debug symbols (so the savings are perhaps just 300 MB).

**Comment by kronbichler** [comment]
Still, this is a nice improvement. :+1:

**Comment by drwells** [comment]
Thanks. I am happy with the final form of this patch.

My only worry is that the number of `#include`s in each unity file is determined manually, as is the separation in each directory between 'big' and 'small' files. I wish there were a way to do these things automatically instead of relying on profiling.

**Comment by bangerth** [comment]
Do you know how cotire does this subdivision?

**Comment by drwells** [comment]
My impression (@tamiko is the expert) is that cotire concatenates all source files for each CMake target into one giant file.

**Comment by drwells** [comment]
Perhaps I should clarify: the problem with concatenating everything is that some of our targets are much more expensive than others. For example, most files in `numerics/` take 30-60 seconds to compile, but most files in `base/` take 2-5 seconds. Hence, for performance, it makes sense to compile most of the `VectorTools` instantiation files separately since they cost about as much as compiling everything in `base` combined.

`fe` is a good example of a directory where some files are huge and some are small: the `FEValues` instantiation files both take about 60 seconds to compile, but most of the finite element classes take about 5-10 seconds. Hence, if we want to aim for targets that each take about 60 seconds to compile, we have to sort things out manually.

**Comment by bangerth** [comment]
cmake stores the information how long it takes each file to compile but I don't know whether it makes that information available from within cmake -- and in any case, that may not be available since it would require that one first does a regular build before a unity build.

How about randomization? Throw random files together and hope for the best.

By the way, how do you create as many subdivisions as there is parallelism based on `-jN`?

**Comment by drwells** [comment]
I don't think that doing things randomly would improve things. Examples: the performance is a lot worse if we don't segregate the PETSc and Trilinos compilation files in `lac/`, and the same is true to a lesser extent for splitting the Mapping classes across multiple unity files.

The number of unity files is not related to the amount of parallelism available: I more-or-less (see exceptions above) added more inclusions to each unity file until each unity target took about 2 GB of memory to compile. We still have O(100) targets in the unity build.

File: source/grid/CMakeLists.txt [hunk 1]
```diff
@@ -15,27 +15,49 @@
...
+    ${_src}
+    ${_output_src}
+    ${_separate_src}
+    )
+ELSE()
+  SET(_src
+    ${_unity_include_src}
+    ${_separate_src}
+    )
+ENDIF()
```
Comment (tamiko) [inline source/grid/CMakeLists.txt]: Just one nit-picky comment. :-)

Can we make this common idiom of 14 lines a simple macro?
Or reduce the block to 6 lines?

**Comment by tamiko** [comment]
I am (almost) happy with the current state of this PR and suggest we merge soon.

@bangerth We (meaning @drwells and me) had a lengthy discussion about unity builds and precompiled headers. We came to the conclusion that it is better to do a manual split of source files by hand than to rely on cotire. Mainly because cotire is a _gigantic_ external script of more than 4000 lines of code (that is of the size of our whole build system) - and we would still need to partition into compilation units by hand...

Further, the precompiled headers part doesn't seem to be very useful (at least for the compilation of deal.II itself).

File: doc/readme.html [hunk 2]
```diff
@@ -398,8 +399,21 @@
       options.
     </p>
 
+    <a name="optional-build"/>
+      <h4>Selecting optional compilation features</h4>
```
Comment (bangerth) [inline doc/readme.html]: in the ToC, you called this "Selecting optional *build* features". Want to settle on one or the other?

**Comment by bangerth** [comment]
I'm not opposed to it. If the two of you think it's a good idea, I'd say let's go with it. Or we can discuss in more detail at the hackathon -- your choice.

File: source/grid/CMakeLists.txt (hunk 1 above)
Comment (drwells) [inline source/grid/CMakeLists.txt]: I'll write another macro here.

**Comment by tjhei** [comment]
> instantiation files

Are you combining any instantiation files? I am asking because I spent quite some time hacking together the split_instantiation macros so we would get memory usage down by splitting into smaller .cc files. It seems this change is working in the opposite direction, so I am worried we might run into issues.

**Comment by kronbichler** [comment]
> Are you combining any instantiation files?

No, as far as I could see the files are all really small ones, so it is not really going in the opposite direction, more like concentrating everything in the middle... ;-)

**Comment by drwells** [comment]
@tjhei Good question: I double checked and none of the files that use split instantiations are concatenated into the unity files. As @kronbichler mentioned I only concatenated files that are cheap to compile separately.

In fact, the only two files in `lac/` that are not put in a unity file are `sparse_matrix.cc` and `sparse_matrix_inst2.cc`. The story is similar for `fe/`.

I aimed to keep the maximum memory required to compile each unity file at about 2 GB so this should not change the maximum amount of memory required.

File: doc/readme.html (hunk 2 above)
Comment (drwells) [inline doc/readme.html]: Thanks for noticing. I went with 'compilation'.

**Comment by drwells** [comment]
I summarized the discussion here in some extra documentation and made the requested macro changes.

**Comment by tamiko** [comment]
@bangerth If you configure with ninja, just have a look at `.ninja_log` first row is start time (in ms after ninja invocation) second row is end time (in ms after ninja invocation), fourth line is the target.

**Comment by bangerth** [comment]
Yes, that's the place I remember. I don't know whether these numbers are accessible from within cmake, though. (Or, more precisely, from within either a `ninja` or `make` file created by `cmake`.)

**Comment by drwells** [comment]
I added another commit where I updated the discussion on sublibrary-level CMakeLists.txt files.

**Comment by drwells** [comment]
@tamiko I would appreciate it if you could double check the most recent commit (given that you are the primary author of the build system).

File: doc/developers/cmake-internals.html [hunk 3]
```diff
@@ -604,6 +605,40 @@
...
+  require about 30 to 60 seconds of wall time and about 2 GB of memory.
+</p>
+
+<p>
+  If you want to add a new file to one of the <code>CMakeLists.txt</code> files
+  then you should place it in either the <code>_unity_include_src</code> or
+  the <code>_separate_src</code> list (not both). If the new file only takes a
+  few seconds to compile then it should be placed in the former category (so
+  that it may be built in a unity file) and otherwise it should be in the
+  later. If you are not sure where a new file belongs then placing it in
```
Comment (kronbichler) [inline doc/developers/cmake-internals.html]: later -> latter

**Comment by tamiko** [comment]
@bangerth No, it is not accessible via cmake [1] - this is purely a ninja specific thing.

[1] Well, if you brutally override cmake internals when generating targets (or use a compiler wrapper that records timings) then technically "yes".

**Comment by tamiko** [comment]
@drwells I approve that latest changes as well. Please address all remaining comments, rebase and merge.

File: doc/developers/cmake-internals.html (hunk 3 above)
Comment (drwells) [inline doc/developers/cmake-internals.html]: Fixed (via amend).

**Comment by drwells** [comment]
I still have not found an acceptable alternative to the compiler wrapping and postprocessing approach I used in #4392: there doesn't seem to be any built-in way to measure memory usage (which was the primary criterion I used in splitting up the unity files) of individual build targets.

**Comment by drwells** [comment]
@tamiko Dumb question: what precisely do you mean by 'rebase'? There are no merge conflicts with master. Would you prefer that the last two commits be squashed together?

**Comment by tamiko** [comment]
> @tamiko Dumb question: what precisely do you mean by 'rebase'? There
> are no merge conflicts with master. Would you prefer that the last two
> commits be squashed together?

No. I would simply prefer that this branch is rebased onto master such
that it doesn't start a month ago.

**Comment by drwells** [comment]
Ah, I see. Yes, these commits have been rebased onto my local branch that is up to date with `master` (see https://github.com/drwells/dealii/commits/manual-unity).

**Pull request merged**